docs/snapshotting/design.md

```diff
@@ -0,0 +1,211 @@
+
+**Return value:** will report success if the snapshot files have been successfully generated, failure otherwise.
+**Body:** Yaml definition: same `SnapshotCreateConfig` body as used in `PUT /snapshot/create`.
+
+### **`PUT /snapshot/load`: will attempt to load a full snapshot of the microVM**
+
+**Prerequisites**:
+- Does **not** work with **diff** snapshots.
+- Can be called only before configuring any resources (other than the `Logger`) for this microVM, or in other words, should be called in a fresh Firecracker process.
+
+**Effects:**
+- _on success_:
+  - Complete microVM state will be loaded from snapshot into this Firecracker process.
+  - Will also reset the dirtied page bitmap, will make all pages *clean* (from a incremental snapshot point of view).
+  - Loaded microVM is now in the *Paused* state and needs to be resumed for it to run.
+  - *The guest-mem snapshot file* from which this microVM was resumed **must not change** throughout the lifetime of this microVM. Since the guest in this microVM will load pages from the file on demand, this Firecracker process should be considered to have a shared-read-lock on the file. Any external writes to the file can lead to microVM guest *Undefined Behavior.*
+  - *The vmm snapshot file* used to load from, is released and no longer used by this process.
+- _on failure_: will report the encountered error, then end this Firecracker process (as it might be in an invalid state).
+
+**Return value:** Will report success if the microVM was successfully loaded, failure otherwise.
+**Body:**
+```yaml
+SnapshotLoadConfig:
+  type: object
+  required:
+    - snapshot_path
+    - mem_file_path
+  properties:
+    snapshot_path:
+      type: string
+      description: Path to file which contains the vmm state to be loaded.
+    mem_file_path:
+      type: string
+      description: Path to file which contains the guest memory to be loaded.
+                   This file should not be externally changed for the entire
+                   lifetime of the microVM to be loaded.
+```
+
+## Snapshot Format
+
+We will use **serde** as a serialization framework since it is a well maintained project with a large community. Serde defines a set of primitives and provides an interface for serializers to implement. Our approach is to build a custom serde serializer that meets the following criteria:
+
+* Performance: Serialization/deserialization should be fast and the snapshot small in size.
+* Code should be well-tested and documented.
+* Implementing cross-version snapshot restore should be easy to do.
+* The format should follow a clearly defined specification.
+* The format should have support in other languages, so tooling can be built around the snapshot format.
+* There should either be a large community that provides good support, or we should own it.
+
+### Current favorites
+
+1. Extend the existing serde compliant **bincode** serializer to match our requirements.
```

Comment:
bincode is interesting, compact, and works well over the network.

However, I think it fails a few of your (sensible) above requirements: it's not very well specified and is quite bound to the rust language.

Reply:
Hi @sameo . I agree that it does not implement a clearly defined spec, but the actual implementation is very simple and easy to understand, extend and improve upon.

We are aware that is currently bound to the rust language, but at this point there are very few usecases defined (updating block device path) that would benefit from the ability to easily edit the snapshot file from outside of Firecracker. Even if we use bincode as the snapshot format, we can support exporting/exporting it as JSON or provide a tool that can handle all future usecases.

Reply:
Sounds very good.